### PR TITLE
Lock incoming nodes from execution context

### PR DESCRIPTION
**Pull request by Mats-SX**
CREATE UNIQUE was producing exceptions in certain concurrent scenarios, where
the same relationship was created twice by different transactions. This was due
to not locking nodes read by previous operators.

### CODE REVIEW

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/CreateUniqueConcurrencyIT.scala [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        counter = counter + errorsFound
+      }
+      execute("MATCH (n) DETACH DELETE n")
+    }
+    counter shouldBe 0
+  }
+
+  def runConcurrentlyToFindDeadlocks(nbrOfNodes: Int, nbrOfThreads: Int): Int = {
+    val random = new Random()
+    val counter = new AtomicInteger(0)
```
Comment (systay) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/CreateUniqueConcurrencyIT.scala:49]: Why is this variable here? It's use on line 75 looks like an assertion started to get written but wasn't finished. I'm guessing that this is to fail the test if any of the threads dies with an exception, right?

File: enterprise/kernel/src/test/scala/org/neo4j/internal/cypher/acceptance/CreateUniqueEnterpriseConcurrencyIT.scala [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        counter = counter + errorsFound
+      }
+      execute("MATCH (n) DETACH DELETE n")
+    }
+    counter shouldBe 0
+  }
+
+  def runConcurrentlyToFindDeadlocks(nbrOfNodes: Int, nbrOfThreads: Int): Int = {
+    val random = new Random()
+    val counter = new AtomicInteger(0)
```
Comment (systay) [inline enterprise/kernel/src/test/scala/org/neo4j/internal/cypher/acceptance/CreateUniqueEnterpriseConcurrencyIT.scala:50]: Same as above

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/CreateUniqueConcurrencyIT.scala (hunk 1 above)
Comment (Mats-SX) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/CreateUniqueConcurrencyIT.scala:49]: The counter's value is returned at the end of the function. The return value for the function is stored in `errorsFound` on line 37, and it is summed up on line 40 in the `counter` defined at line 30, which has the assertion on it on line 44. I think this works as intended.

File: enterprise/kernel/src/test/scala/org/neo4j/internal/cypher/acceptance/CreateUniqueEnterpriseConcurrencyIT.scala (hunk 2 above)
Comment (Mats-SX) [inline enterprise/kernel/src/test/scala/org/neo4j/internal/cypher/acceptance/CreateUniqueEnterpriseConcurrencyIT.scala:50]: The reply is the same too :)

**Pull request merged**